File: bin/ruby-cli/bin/spotlight-manage.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require_relative '../lib/archive/script_base'

# Spotlight Indexing Management Script
# Comprehensive management of macOS Spotlight indexing with privacy focus
class SpotlightManager < ScriptBase
  def banner_text
    <<~BANNER
      🔍 Spotlight Indexing Manager

      Manage macOS Spotlight indexing settings for privacy and performance.
      Control what Spotlight indexes and exclude sensitive directories.

      Usage: #{script_name} [OPTIONS] [COMMAND]
    BANNER
  end

  def add_custom_options(opts)
    opts.on('-s', '--status', 'Show current Spotlight status and settings') do
      @options[:command] = :status
    end

    opts.on('-e', '--enable', 'Enable Spotlight indexing globally') do
      @options[:command] = :enable
    end

    opts.on('-d', '--disable', 'Disable Spotlight indexing globally') do
      @options[:command] = :disable
    end

    opts.on('-r', '--rebuild', 'Rebuild Spotlight index (requires sudo)') do
      @options[:command] = :rebuild
    end

    opts.on('-x', '--exclude PATH', 'Add directory to Spotlight exclusions') do |path|
      @options[:command] = :exclude
      @options[:path] = File.expand_path(path)
    end

    opts.on('-i', '--include PATH', 'Remove directory from Spotlight exclusions') do |path|
      @options[:command] = :include
      @options[:path] = File.expand_path(path)
    end

    opts.on('-p', '--privacy', 'Apply privacy-focused exclusions (recommended)') do
      @options[:command] = :privacy
    end

    opts.on('-l', '--list', 'List current Spotlight exclusions') do
      @options[:command] = :list
    end
  end

  def validate!
    @options[:command] ||= :status

    if [:exclude, :include].include?(@options[:command]) && !@options[:path]
      raise "Path is required for #{@options[:command]} command"
    end

    if @options[:path] && !File.exist?(@options[:path])
      raise "Path does not exist: #{@options[:path]}"
    end

    super
  end

  def run
    log_banner("Spotlight Indexing Manager")

    case @options[:command]
    when :status
      show_status
    when :enable
      enable_spotlight
    when :disable
      disable_spotlight
    when :rebuild
      rebuild_index
    when :exclude
      exclude_path(@options[:path])
    when :include
      include_path(@options[:path])
    when :privacy
      apply_privacy_exclusions
    when :list
      list_exclusions
    else
      log_error("Unknown command: #{@options[:command]}")
      exit 1
    end

    show_completion("Spotlight management")
  end

  private

  def show_status
    log_section("Spotlight Status")

    # Check if Spotlight is enabled
    status = System.execute("mdutil -s /", description: "Checking Spotlight status")
    if status&.include?("Indexing enabled")
      log_success("Spotlight indexing is ENABLED")
    else
      log_warning("Spotlight indexing is DISABLED")
    end

    # Show indexing status for all volumes
    log_info("Checking all volumes...")
    volumes = System.execute("df -h | grep '^/' | awk '{print $9}'",
                             description: "Getting mounted volumes")&.split("\n") || []

    volumes.each do |volume|
      next if volume.empty?

      volume_status = System.execute("mdutil -s '#{volume}'", description: "Checking #{volume}")
      if volume_status.include?("Indexing enabled")
        log_info("✅ #{volume}: Enabled")
      else
        log_info("❌ #{volume}: Disabled")
      end
    end

    # Show current exclusions count
    exclusions = get_spotlight_exclusions
    log_info("Current exclusions: #{exclusions.count} directories")

    # Show indexing progress if active
    log_info("Checking indexing progress...")
    progress = System.execute("mdutil -t /", description: "Checking indexing progress")
    unless progress.empty?
      log_info("Indexing progress: #{progress}")
    end
  end

  def enable_spotlight
    log_section("Enabling Spotlight")

    if confirm_action("Enable Spotlight indexing on all volumes?")
      log_progress("Enabling Spotlight indexing...")

      # Enable for root volume
      System.execute("sudo mdutil -i on /", description: "Enabling Spotlight on root volume")

      # Enable for all mounted volumes
      volumes = System.execute("df -h | grep '^/' | awk '{print $9}'",
                               description: "Getting mounted volumes").split("\n")
      volumes.each do |volume|
        next if volume.empty? || volume == "/"

        System.execute("sudo mdutil -i on '#{volume}'", description: "Enabling Spotlight on #{volume}")
      end

      log_success("Spotlight indexing enabled")
      log_info("Indexing will start automatically and may take some time to complete")
    else
      log_info("Operation cancelled")
    end
  end

  def disable_spotlight
    log_section("Disabling Spotlight")

    log_warning("Disabling Spotlight will:")
    log_warning("• Stop all indexing processes")
    log_warning("• Make Spotlight search unavailable")
    log_warning("• Affect apps that rely on Spotlight (like Alfred, Raycast)")

    if confirm_action("Are you sure you want to disable Spotlight?")
      log_progress("Disabling Spotlight indexing...")

      # Disable for root volume
      System.execute("sudo mdutil -i off /", description: "Disabling Spotlight on root volume")

      # Disable for all mounted volumes
      volumes = System.execute("df -h | grep '^/' | awk '{print $9}'",
                               description: "Getting mounted volumes").split("\n")
      volumes.each do |volume|
        next if volume.empty? || volume == "/"

        System.execute("sudo mdutil -i off '#{volume}'", description: "Disabling Spotlight on #{volume}")
      end

      log_success("Spotlight indexing disabled")
    else
      log_info("Operation cancelled")
    end
  end

  def rebuild_index
    log_section("Rebuilding Spotlight Index")

    log_warning("Rebuilding the index will:")
    log_warning("• Delete the current Spotlight index")
    log_warning("• Start a fresh indexing process")
    log_warning("• Take significant time to complete")
    log_warning("• Use CPU and disk resources during indexing")

    if confirm_action("Rebuild Spotlight index?")
      log_progress("Rebuilding Spotlight index...")

      # Erase and rebuild index
      System.execute("sudo mdutil -E /", description: "Erasing and rebuilding Spotlight index")

      log_success("Spotlight index rebuild initiated")
      log_info("Indexing will start automatically and may take 30+ minutes to complete")
      log_info("You can check progress with: mdutil -t /")
    else
      log_info("Operation cancelled")
    end
  end

  def exclude_path(path)
    log_section("Adding Spotlight Exclusion")

    log_info("Adding to exclusions: #{path}")

    if get_spotlight_exclusions.include?(path)
      log_warning("Path is already excluded from Spotlight")
      return
    end

    # Use Spotlight preferences to add exclusion
    System.execute("sudo mdutil -X '#{path}'", description: "Adding Spotlight exclusion")

    # Alternative method using defaults (more reliable)
    exclusions = get_spotlight_exclusions
    exclusions << path unless exclusions.include?(path)

    # Convert to the format expected by Spotlight
    exclusion_data = exclusions.map { |p| "<string>#{p}</string>" }.join("\n")

    if @options[:dry_run]
      log_info("Would add exclusion: #{path}")
    else
      log_success("Added Spotlight exclusion: #{path}")
      log_info("Changes may take a few minutes to take effect")
    end
  end

  def include_path(path)
    log_section("Removing Spotlight Exclusion")

    log_info("Removing from exclusions: #{path}")

    exclusions = get_spotlight_exclusions
    unless exclusions.include?(path)
      log_warning("Path is not currently excluded from Spotlight")
      return
    end

    if @options[:dry_run]
      log_info("Would remove exclusion: #{path}")
    else
      # Remove the exclusion (this is complex with macOS preferences)
      log_warning("Manual removal required:")
      log_info("1. Open System Preferences > Spotlight > Privacy")
      log_info("2. Find and remove: #{path}")
      log_info("3. Or use: sudo mdutil -R '#{path}'")
    end
  end

  def apply_privacy_exclusions
    log_section("Applying Privacy-Focused Exclusions")

    privacy_paths = [
      "#{Dir.home}/Downloads",
      "#{Dir.home}/Documents/Private",
      "#{Dir.home}/.ssh",
      "#{Dir.home}/.gnupg",
      "#{Dir.home}/Library/Keychains",
      "#{Dir.home}/Library/Application Support/1Password",
      "#{Dir.home}/Library/Cookies",
      "#{Dir.home}/.config",
      "/usr/local/var",
      "/opt/homebrew/var",
      "#{Dir.home}/.cache",
      "#{Dir.home}/.local",
      "#{Dir.home}/node_modules",
      "#{Dir.home}/Library/Caches"
    ].select { |path| File.exist?(path) }

    log_info("Privacy paths to exclude:")
    privacy_paths.each { |path| log_info("  • #{path}") }

    if confirm_action("Apply these privacy exclusions?")
      privacy_paths.each do |path|
        log_progress("Excluding #{File.basename(path)}...")
        exclude_path(path) unless get_spotlight_exclusions.include?(path)
      end

      log_success("Privacy exclusions applied")
    else
      log_info("Operation cancelled")
    end
  end

  def list_exclusions
    log_section("Current Spotlight Exclusions")

    exclusions = get_spotlight_exclusions

    if exclusions.empty?
      log_info("No directories are currently excluded from Spotlight")
    else
      log_info("Excluded directories (#{exclusions.count}):")
      exclusions.each_with_index do |path, index|
        status = File.exist?(path) ? "✅" : "❌ (missing)"
        log_info("#{(index + 1).to_s.rjust(2)}. #{path} #{status}")
      end
    end
  end

  def get_spotlight_exclusions
    # Read Spotlight preferences to get current exclusions
    prefs_cmd = "defaults read com.apple.Spotlight orderedItems 2>/dev/null || echo '()'"
    prefs_output = System.execute(prefs_cmd, description: "Reading Spotlight preferences")

    # This is a simplified version - actual implementation would parse the plist properly
    exclusions = []

    # Try to get exclusions from mdutil
    volumes = ["/"]
    volumes.each do |volume|
      begin
        mdutil_output = System.execute("mdutil -s #{volume}", description: "Checking exclusions")
        # Parse exclusions from output (implementation depends on mdutil format)
      rescue
        # Handle errors gracefully
      end
    end

    # Add some common exclusions that are typically present
    common_exclusions = [
      "#{Dir.home}/.Trash",
      "/private/var/vm"
    ].select { |path| File.exist?(path) }

    (exclusions + common_exclusions).uniq.sort
  end

  def show_examples
    puts <<~EXAMPLES

      Examples:
        #{script_name} --status                    # Show current Spotlight status
        #{script_name} --disable                   # Disable Spotlight indexing
        #{script_name} --enable                    # Enable Spotlight indexing
        #{script_name} --rebuild                   # Rebuild the index
        #{script_name} --exclude ~/Private         # Exclude directory
        #{script_name} --include ~/Downloads       # Remove exclusion
        #{script_name} --privacy                   # Apply privacy exclusions
        #{script_name} --list                      # List current exclusions
        #{script_name} --dry-run --privacy         # Preview privacy changes
    EXAMPLES
  end
end

# Execute the script
SpotlightManager.execute if __FILE__ == $0

File: bin/bun-cli/src/scripts/macos/spotlight-manage.ts
import { Script } from '../../core/decorators/Script';
import { Script as BaseScript } from '../../core/base/Script';
import type { Context } from '../../core/types';
import { SystemService } from '../../core/services/SystemService';
import { execSilent } from '../../core/utils/shell';
import * as clack from '@clack/prompts';

interface VolumeInfo {
  path: string;
  indexing: 'Enabled' | 'Disabled';
  status?: string;
}

/**
 * Manage macOS Spotlight indexing
 *
 * Control Spotlight indexing on volumes, rebuild indices, and manage
 * privacy exclusions for improved performance and privacy.
 *
 * @example
 * spotlight-manage
 * spotlight-manage --status
 * spotlight-manage --disable /Volumes/External
 * spotlight-manage --rebuild
 */
@Script({
  emoji: '🔍',
  tags: ['macos', 'spotlight', 'privacy'],
  args: {
    status: {
      type: 'boolean',
      flag: '--status',
      description: 'Show current Spotlight indexing status',
    },
    enable: {
      type: 'string',
      flag: '--enable <volume>',
      description: 'Enable Spotlight indexing on volume',
    },
    disable: {
      type: 'string',
      flag: '--disable <volume>',
      description: 'Disable Spotlight indexing on volume',
    },
    rebuild: {
      type: 'boolean',
      flag: '--rebuild',
      description: 'Rebuild Spotlight index',
    },
    privacy: {
      type: 'boolean',
      flag: '--privacy',
      description: 'Add common privacy exclusions',
    },
  },
})
export class SpotlightManageScript extends BaseScript {
  private readonly commonExclusions = [
    '~/Library/Application Support/com.apple.TCC',
    '~/Library/Application Support/Google/Chrome',
    '~/Library/Application Support/Slack',
    '~/Library/Caches',
    '~/Library/Logs',
    '~/.Trash',
    '~/Downloads',
    '~/node_modules',
    '~/.npm',
    '~/.cache',
    '~/.docker',
  ];

  async validate(ctx: Context): Promise<void> {
    // Validate macOS
    if (!SystemService.isMacOS()) {
      throw new Error('This script requires macOS');
    }

    // Check for required commands
    if (!(await SystemService.commandExists('mdutil'))) {
      throw new Error('Required command not found: mdutil');
    }
  }

  async run(ctx: Context): Promise<void> {
    const { status, enable, disable, rebuild, privacy } = ctx.args;

    this.logger.section('Spotlight Management');

    // Handle specific actions
    if (status) {
      await this.showStatus();
      return;
    }

    if (enable) {
      await this.enableIndexing(enable);
      return;
    }

    if (disable) {
      await this.disableIndexing(disable);
      return;
    }

    if (rebuild) {
      await this.rebuildIndex();
      return;
    }

    if (privacy) {
      await this.addPrivacyExclusions();
      return;
    }

    // Interactive mode
    await this.interactiveMode();
  }

  private async interactiveMode(): Promise<void> {
    const action = await clack.select({
      message: 'What would you like to do?',
      options: [
        { value: 'status', label: 'Show Spotlight status' },
        { value: 'enable', label: 'Enable indexing on volume' },
        { value: 'disable', label: 'Disable indexing on volume' },
        { value: 'rebuild', label: 'Rebuild Spotlight index' },
        { value: 'privacy', label: 'Add privacy exclusions' },
        { value: 'exit', label: 'Exit' },
      ],
    });

    if (clack.isCancel(action) || action === 'exit') {
      this.logger.info('Exiting');
      return;
    }

    switch (action) {
      case 'status':
        await this.showStatus();
        break;
      case 'enable':
        await this.enableIndexingInteractive();
        break;
      case 'disable':
        await this.disableIndexingInteractive();
        break;
      case 'rebuild':
        await this.rebuildIndex();
        break;
      case 'privacy':
        await this.addPrivacyExclusions();
        break;
    }
  }

  private async showStatus(): Promise<void> {
    this.logger.info('Checking Spotlight status on all volumes...');

    const volumes = await this.getVolumes();

    if (volumes.length === 0) {
      this.logger.warn('No volumes found');
      return;
    }

    console.log();
    console.log('Spotlight Indexing Status:');
    console.log('─'.repeat(60));

    for (const volume of volumes) {
      const statusIcon = volume.indexing === 'Enabled' ? '✅' : '❌';
      console.log(`${statusIcon} ${volume.path}`);
      console.log(`   Indexing: ${volume.indexing}`);
      if (volume.status) {
        console.log(`   Status: ${volume.status}`);
      }
      console.log();
    }

    this.logger.success('Spotlight status check complete');
  }

  private async getVolumes(): Promise<VolumeInfo[]> {
    try {
      const output = await execSilent('mdutil -a -s');
      const volumes: VolumeInfo[] = [];
      const lines = output.split('\n').filter((line) => line.trim());

      let currentVolume: Partial<VolumeInfo> | null = null;

      for (const line of lines) {
        if (line.includes(':')) {
          // Save previous volume if exists
          if (currentVolume && currentVolume.path && currentVolume.indexing) {
            volumes.push(currentVolume as VolumeInfo);
          }

          // Start new volume
          currentVolume = {
            path: line.replace(':', '').trim(),
          };
        } else if (currentVolume && line.includes('Indexing')) {
          const match = line.match(/Indexing\s+(enabled|disabled)/i);
          if (match) {
            currentVolume.indexing = match[1].charAt(0).toUpperCase() + match[1].slice(1).toLowerCase() as 'Enabled' | 'Disabled';
          }
        } else if (currentVolume && line.includes('Status:')) {
          currentVolume.status = line.replace('Status:', '').trim();
        }
      }

      // Save last volume
      if (currentVolume && currentVolume.path && currentVolume.indexing) {
        volumes.push(currentVolume as VolumeInfo);
      }

      return volumes;
    } catch (error: any) {
      this.logger.error(`Failed to get volumes: ${error.message}`);
      return [];
    }
  }

  private async enableIndexing(volumePath: string): Promise<void> {
    this.logger.info(`Enabling Spotlight indexing on ${volumePath}...`);

    try {
      const output = await execSilent(`sudo mdutil -i on "${volumePath}"`);
      this.logger.success(`Indexing enabled on ${volumePath}`);

      if (output.includes('will continue')) {
        this.logger.info('Index will be rebuilt automatically');
      }
    } catch (error: any) {
      this.logger.error(`Failed to enable indexing: ${error.message}`);
      process.exit(1);
    }
  }

  private async disableIndexing(volumePath: string): Promise<void> {
    this.logger.info(`Disabling Spotlight indexing on ${volumePath}...`);

    try {
      await execSilent(`sudo mdutil -i off "${volumePath}"`);
      this.logger.success(`Indexing disabled on ${volumePath}`);
    } catch (error: any) {
      this.logger.error(`Failed to disable indexing: ${error.message}`);
      process.exit(1);
    }
  }

  private async enableIndexingInteractive(): Promise<void> {
    const volumes = await this.getVolumes();
    const disabledVolumes = volumes.filter((v) => v.indexing === 'Disabled');

    if (disabledVolumes.length === 0) {
      this.logger.info('All volumes already have indexing enabled');
      return;
    }

    const selected = await clack.select({
      message: 'Select volume to enable indexing:',
      options: disabledVolumes.map((v) => ({
        value: v.path,
        label: v.path,
      })),
    });

    if (clack.isCancel(selected)) {
      this.logger.info('Operation cancelled');
      return;
    }

    await this.enableIndexing(selected as string);
  }

  private async disableIndexingInteractive(): Promise<void> {
    const volumes = await this.getVolumes();
    const enabledVolumes = volumes.filter((v) => v.indexing === 'Enabled');

    if (enabledVolumes.length === 0) {
      this.logger.info('All volumes already have indexing disabled');
      return;
    }

    const selected = await clack.select({
      message: 'Select volume to disable indexing:',
      options: enabledVolumes.map((v) => ({
        value: v.path,
        label: v.path,
      })),
    });

    if (clack.isCancel(selected)) {
      this.logger.info('Operation cancelled');
      return;
    }

    const selectedPath = selected as string;

    const confirm = await clack.confirm({
      message: `Disable Spotlight indexing on ${selectedPath}?`,
    });

    if (clack.isCancel(confirm) || !confirm) {
      this.logger.info('Operation cancelled');
      return;
    }

    await this.disableIndexing(selectedPath);
  }

  private async rebuildIndex(): Promise<void> {
    const volumes = await this.getVolumes();
    const enabledVolumes = volumes.filter((v) => v.indexing === 'Enabled');

    if (enabledVolumes.length === 0) {
      this.logger.warn('No volumes with indexing enabled');
      return;
    }

    let volumePath: string;

    if (enabledVolumes.length === 1) {
      volumePath = enabledVolumes[0].path;
    } else {
      const selected = await clack.select({
        message: 'Select volume to rebuild index:',
        options: enabledVolumes.map((v) => ({
          value: v.path,
          label: v.path,
        })),
      });

      if (clack.isCancel(selected)) {
        this.logger.info('Operation cancelled');
        return;
      }

      volumePath = selected as string;
    }

    const confirm = await clack.confirm({
      message: `Rebuild Spotlight index on ${volumePath}? This may take some time.`,
    });

    if (clack.isCancel(confirm) || !confirm) {
      this.logger.info('Operation cancelled');
      return;
    }

    this.logger.info(`Erasing current index on ${volumePath}...`);

    try {
      await execSilent(`sudo mdutil -E "${volumePath}"`);
      this.logger.success('Index erased');
      this.logger.info('Spotlight will rebuild the index automatically');
      this.logger.info('This process may take several minutes to hours depending on volume size');
    } catch (error: any) {
      this.logger.error(`Failed to rebuild index: ${error.message}`);
      process.exit(1);
    }
  }

  private async addPrivacyExclusions(): Promise<void> {
    this.logger.section('Privacy Exclusions');
    this.logger.info('Adding common privacy exclusions to Spotlight...');

    const homeDir = SystemService.getHomeDir();
    const exclusionsToAdd: string[] = [];

    // Expand ~ to home directory and check if paths exist
    for (const exclusion of this.commonExclusions) {
      const expandedPath = exclusion.replace('~', homeDir);

      if (await this.fs.exists(expandedPath)) {
        exclusionsToAdd.push(expandedPath);
      }
    }

    if (exclusionsToAdd.length === 0) {
      this.logger.warn('No common exclusion paths found on this system');
      return;
    }

    console.log();
    console.log('Paths to exclude from Spotlight:');
    for (const path of exclusionsToAdd) {
      console.log(`  • ${path}`);
    }
    console.log();

    const confirm = await clack.confirm({
      message: `Add ${exclusionsToAdd.length} privacy exclusions?`,
    });

    if (clack.isCancel(confirm) || !confirm) {
      this.logger.info('Operation cancelled');
      return;
    }

    this.logger.info('Adding exclusions...');
    this.logger.warn(
      'Note: You will need to manually add these to System Settings > Spotlight > Privacy'
    );
    this.logger.info('This script will open System Settings for you');

    console.log();
    console.log('Copy these paths and add them manually:');
    console.log('─'.repeat(60));
    for (const path of exclusionsToAdd) {
      console.log(path);
    }
    console.log('─'.repeat(60));

    const shouldOpen = await clack.confirm({
      message: 'Open System Settings > Spotlight > Privacy?',
    });

    if (!clack.isCancel(shouldOpen) && shouldOpen) {
      try {
        await execSilent('open "x-apple.systempreferences:com.apple.Spotlight-Settings.extension"');
        this.logger.success('System Settings opened');
      } catch (error: any) {
        this.logger.error(`Failed to open System Settings: ${error.message}`);
      }
    }

    this.logger.success('Privacy exclusions ready to add');
  }
}
